DatadogSessionReplay/Sources/Recorder/RecordingCoordinator.swift

```diff
@@ -115,46 +113,68 @@ internal class RecordingCoordinator {
         srContextPublisher.setHasReplay(hasReplay)
     }
 
-    /// Captures the next recording if conditions are met.
-    private func captureNextRecord() {
-        /// We don't capture any snapshots if the RUM context has no view ID.
-        guard let rumContext = currentRUMContext,
-              let viewID = rumContext.viewID else {
-            return
-        }
+    private var lastFrameTimestamp: Date?
+    private var shouldSkipFrame: Bool {
+        return Date().timeIntervalSince(lastFrameTimestamp ?? .distantPast) < Constants.throttingRate
+    }
 
-        let recorderContext = Recorder.Context(
-            textAndInputPrivacy: textAndInputPrivacy,
-            imagePrivacy: imagePrivacy,
-            touchPrivacy: touchPrivacy,
-            applicationID: rumContext.applicationID,
-            sessionID: rumContext.sessionID,
-            viewID: viewID,
-            viewServerTimeOffset: rumContext.viewServerTimeOffset
-        )
-
-        let methodCalledTrace = telemetry.startMethodCalled(
-            operationName: MethodCallConstants.captureRecordOperationName,
-            callerClass: MethodCallConstants.className,
-            headSampleRate: methodCallTelemetrySamplingRate // Effectively 3% * 0.1% = 0.003% of calls
-        )
-
-        var isSuccessful = false
-        do {
-            try objc_rethrow { try recorder.captureNextRecord(recorderContext) }
-            isSuccessful = true
-        } catch let objc as ObjcException {
-            telemetry.error("[SR] Failed to take snapshot due to Objective-C runtime exception", error: objc.error)
-            // An Objective-C runtime exception is a severe issue that will leak if
-            // the framework is not built with `-fobjc-arc-exceptions` option.
-            // We recover from the exception and stop the scheduler as a measure of
-            // caution. The scheduler could start again at a next RUM context change.
-            stopRecording()
-        } catch {
-            telemetry.error("[SR] Failed to take snapshot", error: error)
+    /// Captures the next recording if conditions are met.
+    func captureNextRecord() {
+        queue.run { [weak self] in
+            guard let self = self else {
+                return
+            }
+            guard isSampled == true && recordingEnabled == true else {
+                return
+            }
+            // We don't capture any snapshots if the RUM context has no view ID.
+            guard let rumContext = currentRUMContext,
+                  let viewID = rumContext.viewID else {
+                return
+            }
+            // Skip frame if there's pressure on processing
+            guard shouldSkipFrame == false else {
+                return
+            }
+            lastFrameTimestamp = Date()
```

Comment:
change-request/ Throttling from inside the async operation is quite inefficient as it leads to flooding the `main` queue with bunch of unnecessary tasks. From my test in Shopist app:
- when app is launched, the `main` queue is flooded with nearly 200 async tasks due to `layoutSubviews` trigger;
- when touching the screen and moving finger, it is again flooded with about 50 tasks per second.

It can be avoided if the "should skip frame" check is moved before the task gets dispatched. Note that we discuss the performance impact on the UI thread, so we must do whatever it takes to unload it from unnecessary consumption.